File: TestProject/Entity.cpp
#include "Entity.h"


Entity::Entity(GraphicsContext *context)
	: context(context)
{
}

Entity::~Entity()
{
}

void Entity::getSize(Vec2 *vec2) const
{
	TextureSize size;
	context->getTexture2DSize(image, &size);

	vec2->x = (float)size.width;
	vec2->y = (float)size.height;
}

void Entity::draw(SpriteBatch &spriteBatch)
{
	Vec2 size;
	getSize(&size);

	Sprite sprite;
	sprite.setPositionAnchor(0.5f, 0.5f);
	sprite.setPosition(position.x, position.y);
	sprite.setSize(size.x, size.y);
	sprite.setSrcRect(FloatRect(0.0f, 0.0f, size.x, size.y));
	sprite.setAngle(orientation);
	sprite.setColor(color);
	sprite.setTexture(image);

	spriteBatch.draw(sprite);
}

File: CSTestProject/Entity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

abstract class Entity
{
    protected IntPtr image;
    protected Color color = Color.White;

    public Vector2 Position, Velocity;
    public float Orientation;
    public float Radius = 20;
    public bool IsExpired;

    public Vector2 Size
    {
        get
        {
            if (image == null)
                return Vector2.Zero;

            TextureSize size = GameRoot.Context.GetTexture2DSize(image);
            return new Vector2(size.Width, size.Height);
        }
    }

    public abstract void Update();

    public virtual void Draw(SpriteBatch spriteBatch)
    {
		Sprite sprite = new Sprite();
		sprite.SetPositionAnchor(0.5f, 0.5f);
		sprite.SetPosition(Position.X, Position.Y);
		sprite.SetSize(Size.X, Size.Y);
		sprite.SetSrcRect(0, 0, Size.X, Size.Y);
		sprite.SetAngle(Orientation);
		sprite.SetColor(color);
		sprite.SetTexture(image);

		spriteBatch.Draw(sprite);
	}
}
